File: c1.hs
chunk :: [a] -> Int -> [[a]]
chunk array sizeOfChunks = [mapChunk x | x <- [0 .. numberOfChunks - 1]]
  where
    numberOfElements = length array
    numberOfChunks = ceiling $ fromIntegral numberOfElements / fromIntegral sizeOfChunks
    mapChunk val =
      let startingIndex = val * sizeOfChunks
          endingIndex = startingIndex + sizeOfChunks
       in slice startingIndex endingIndex array

slice :: Int -> Int -> [a] -> [a]
slice from to xs = take (to - from) (drop from xs)

chunk' :: [a] -> Int -> [[a]]
chunk' array sizeOfChunks
  | sizeOfChunks <= 0 || null array = []
  | otherwise =
      let group' = take sizeOfChunks array
          rest' = drop sizeOfChunks array
       in group' : chunk' rest' sizeOfChunks

-- >>> chunk' [1,2,3,4,5,6] 2
-- [[1,2],[3,4],[5,6]]

-- >>> chunk [1,2,3,4] 3
-- [[1,2,3],[4]]

File: c1.js
function chunk(array, sizeOfChunks) {
  const numberOfElements = array.length;
  const numberOfChunks = Math.ceil(numberOfElements / sizeOfChunks);
  const arrayOfChunks = [...new Array(numberOfChunks)].map((_,index) => {
    const startingIndex = index * sizeOfChunks;
    const endingIndex = startingIndex + sizeOfChunks;
    return array.slice(startingIndex,endingIndex);
  });

  return arrayOfChunks;
}

function chunk2(array, sizeOfChunks) {
  if (!array | array.length === 0 || sizeOfChunks <= 0) return [];
  const group = array.slice(0, sizeOfChunks);
  const rest = array.slice(sizeOfChunks);
  return [group].concat(chunk2(rest, sizeOfChunks))
}

console.log(chunk2([1,2,3,4,5,6], 3));